### PR TITLE
[Chore] refactor of ANALYTICS microservice

### PR DESCRIPTION
**Pull request by Baalmart**
#Chore

**_WHAT DOES THIS PR DO?_**

- [x] Adds multitenancy to Analytics microservice 

- [x] Refactors the codebase to incorporate best practices 

- [ ] Rewrite controller logic to utilise the new Events collection. This is related to this [PR for modifying the Events collection](https://github.com/airqo-platform/AirQo-api/pull/254) 

- [x] Separates runtime ENV for the microservice(s) 

- [ ] Creates 3 new microservices/jobs, that is: categories, exceedances and averages 

- [ ] Transfers the ETL functions (update hourly, raw and daily measurements) to an appropriate microservice (data-mgt).

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
[PLAT-240](https://airqoteam.atlassian.net/browse/PLAT-240?atlOrigin=eyJpIjoiODFiYmE5YjE2MDM4NDFmMzg2ZjBkYzJmZTg0NmMxZjYiLCJwIjoiaiJ9)

**_WHAT IS THE LINK TO THE PR BRANCH_**
https://github.com/airqo-platform/AirQo-api/tree/chore-refactor-analytics

**_HOW DO I TEST OUT THIS PR?_**
cd AirQo-api/src/analytics

> For Linux

`python3 -m venv ~/venv`
`source ~/venv/bin/activate`
`pip install -r requirements.txt`
Set your `.env` file. The current staging version is [here](https://airqo.slack.com/archives/C0278FK2EGG/p1625194514000400)
`flask run`

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 
[POST] _Getting the historical daily averages for the first 28 days:_ http://localhost:5000/api/v1/dashboard/historical/daily/devices?tenant=analytics

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
N/A

**_ARE THERE ANY RELATED PRs?_**
In case you experience any challenges with running the app, try to activate your new environment again using `source ~/venv/bin/activate`

### CODE REVIEW

File: src/exceedances/job/views/exceedance.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+        projection = {'$project': {'_id': 0,
+                                   'time': {'$dateToString': {'format': '%Y-%m-%dT%H:%M:%S%z', 'date': '$time', 'timezone': 'Africa/Kampala'}},
+                                   'pollutant_value': {'$round': ['$characteristics.no2Conc.value', 2]}}}
+    else:
+        projection = {'$project': {'_id': 0,
+                                   'time': {'$dateToString': {'format': '%Y-%m-%dT%H:%M:%S%z', 'date': '$time', 'timezone': 'Africa/Kampala'}},
+                                   'pollutant_value': {'$round': ['$characteristics.pm2_5ConcMass.value', 2]}}}
+
+    if frequency == 'hourly':
+        # shall just query from the Events collection accordingly.
```
Comment (Baalmart) [inline src/exceedances/job/views/exceedance.py:157]: This job (exceedances) will just query from the Events collection accordingly.....as one of the collections. @ja-odur

File: src/exceedances/job/views/exceedance.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+                      'characteristics.pm10ConcMass.value': 1}
+        if frequency == 'hourly':
+            records = list(
+                db.device_hourly_measurements.find(query, projection))
+        elif frequency == 'daily':
+            records = list(
+                db.device_daily_measurements.find(query, projection))
+        else:
+            records = list(db.device_raw_measurements.find(query, projection))
+        for i in range(len(records)):
```
Comment (Baalmart) [inline src/exceedances/job/views/exceedance.py:206]: This is where the most important logic is for calculation, where the exceedances are calculated based on some predetermined thresholds, one for WHO and another one just AQI. @ja-odur

File: src/exceedances/job/views/exceedance.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        convert_date.date_to_str(datetime.now()))
+    for pollutant in pollutants:
+        for standard in standards:
+            exceedences_data = get_exceedances(pollutant, standard)
+            exceedence_record = {'pollutant': pollutant, 'standard': standard, 'exceedences': exceedences_data,
+                                 'created_at': created_at}
+            exceedence_results.append(exceedence_record)
+
+    print(exceedence_results)
+    save_device_daily_exceedences(exceedence_results)
```
Comment (Baalmart) [inline src/exceedances/job/views/exceedance.py:291]: So afterwards, these exceedances are stored for future use. @ja-odur

Comment (Baalmart) [reply src/exceedances/job/views/exceedance.py:291]: The purpose of this storage is mainly for optimisation purposes. So the question is: are these values needed for future analysis by the data scientists or we can just cache them and expire them at a predetermined time. @ja-odur

File: src/exceedances/job/views/exceedance.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+    return list(records)
+
+
+def save_device_daily_exceedences(data):
+    """
+    # this ones moves to the exceedances model/entity
+    """
+    for i in data:
+        db.device_daily_exceedences.insert_one(i)
```
Comment (Baalmart) [inline src/exceedances/job/views/exceedance.py:175]: this is the name of the collection which stores the values.  @ja-odur See screenshot....
Screenshot 2021-06-25 at 12 40 39

**Review by Baalmart — COMMENTED** [review]
Hi @ja-odur 
Exceedances job/micro might just need/utilise 2 collections. See rest of comments for more insights.

- **Events** for fetching the measurements
- **Exceeedances** for storage of these exceedances which is done for optimisation purposes I guess.

cc: @danielogen

File: src/exceedances/job/views/exceedance.py (hunk 3 above)
Comment (ja-odur) [reply src/exceedances/job/views/exceedance.py:291]: Yes @Baalmart This has to be considered

File: src/exceedances/job/views/exceedance.py (hunk 4 above)
Comment (ja-odur) [reply src/exceedances/job/views/exceedance.py:175]: Noted. Are gonna maintain the same structure @Baalmart

File: src/data-mgt/python/helpers/insert.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
+import json
+import os
+import requests
+
+AIRQO_API_BASE_URL = os.getenv("AIRQO_API_BASE_URL")
```
Comment (Baalmart) [inline src/data-mgt/python/helpers/insert.py:5]: this code should not be here, we need to review it. cc: @NoahNsimbe

**Pull request merged**